File: obs_replaykit/bongo_cat.py
"""install the Bongobs/Bango Cat OBS plugin from the bundled zip."""

from __future__ import annotations

import shutil
import zipfile
from pathlib import Path, PurePosixPath
from typing import Callable, Optional

from .config import BONGO_CAT_ZIP, PROGRAMFILES_OBS_DIR


LogFn = Optional[Callable[[str], None]]

_ZIP_ROOT = "Bango Cat/"
_PLUGIN_DLL = PROGRAMFILES_OBS_DIR / "obs-plugins" / "64bit" / "bongobs-cat.dll"
_PLUGIN_DATA = PROGRAMFILES_OBS_DIR / "bin" / "64bit" / "Bango Cat"


def _is_safe_relative(path: PurePosixPath) -> bool:
    return bool(path.parts) and all(part not in ("", ".", "..") for part in path.parts)


def _target_path(rel: PurePosixPath) -> Path:
    target_root = PROGRAMFILES_OBS_DIR.resolve(strict=False)
    target = (PROGRAMFILES_OBS_DIR / Path(*rel.parts)).resolve(strict=False)
    if target != target_root and target_root not in target.parents:
        raise ValueError(f"unsafe archive path: {rel}")
    return target


def is_bongo_cat_installed() -> bool:
    return _PLUGIN_DLL.is_file() and _PLUGIN_DATA.is_dir()


def install_bongo_cat_plugin(log: LogFn = None) -> bool:
    """extract the Bango Cat archive into the OBS install root."""
    if is_bongo_cat_installed():
        if log:
            log(f"Bongo Cat plugin already installed at {_PLUGIN_DLL}")
        return True
    if not PROGRAMFILES_OBS_DIR.is_dir():
        if log:
            log(f"OBS install folder was not found: {PROGRAMFILES_OBS_DIR}")
        return False
    if not BONGO_CAT_ZIP.is_file():
        if log:
            log(f"{BONGO_CAT_ZIP.name} is not bundled in assets; skipping Bongo Cat")
        return False

    if log:
        log(f"extracting {BONGO_CAT_ZIP.name} -> {PROGRAMFILES_OBS_DIR}")

    try:
        with zipfile.ZipFile(BONGO_CAT_ZIP) as zf:
            for entry in zf.infolist():
                name = entry.filename.replace("\\", "/")
                if not name.startswith(_ZIP_ROOT):
                    continue
                rel_text = name[len(_ZIP_ROOT):].strip("/")
                if not rel_text:
                    continue
                rel = PurePosixPath(rel_text)
                if not _is_safe_relative(rel):
                    raise ValueError(f"unsafe archive path: {entry.filename}")
                dest = _target_path(rel)
                if entry.is_dir():
                    dest.mkdir(parents=True, exist_ok=True)
                    continue
                dest.parent.mkdir(parents=True, exist_ok=True)
                with zf.open(entry) as src, dest.open("wb") as dst:
                    shutil.copyfileobj(src, dst)
    except Exception as exc:
        if log:
            log(f"Bongo Cat extraction failed: {exc}")
        return False

    if not _PLUGIN_DLL.is_file():
        if log:
            log(f"Bongo Cat extraction finished but {_PLUGIN_DLL} is missing")
        return False
    if not _PLUGIN_DATA.is_dir():
        if log:
            log(f"Bongo Cat extraction finished but {_PLUGIN_DATA} is missing")
        return False

    if log:
        log(f"installed -> {_PLUGIN_DLL}")
        log(f"           + {_PLUGIN_DATA}")
    return True

File: ReplayKitSetup/BongoCat.cs
using System;
using System.IO;
using System.IO.Compression;
using System.Linq;

namespace ReplayKitSetup
{
    // install the bongobs/bango cat obs plugin from the bundled zip. ported from obs_replaykit/bongo_cat.py.
    public static class BongoCat
    {
        private const string ZipRoot = "Bango Cat/";
        private static readonly string PluginDll = Path.Combine(Config.PROGRAMFILES_OBS_DIR, "obs-plugins", "64bit", "bongobs-cat.dll");
        private static readonly string PluginData = Path.Combine(Config.PROGRAMFILES_OBS_DIR, "bin", "64bit", "Bango Cat");

        private static bool IsSafeRelative(string[] parts)
        {
            return parts.Length > 0 && parts.All(p => p != "" && p != "." && p != "..");
        }

        private static string TargetPath(string relPosix)
        {
            string targetRoot = Path.GetFullPath(Config.PROGRAMFILES_OBS_DIR).TrimEnd('\\', '/');
            var parts = relPosix.Split('/');
            string target = Path.GetFullPath(Path.Combine(new[] { Config.PROGRAMFILES_OBS_DIR }.Concat(parts).ToArray()));
            if (!string.Equals(target, targetRoot, StringComparison.OrdinalIgnoreCase) &&
                !target.StartsWith(targetRoot + Path.DirectorySeparatorChar, StringComparison.OrdinalIgnoreCase))
            {
                throw new InvalidOperationException("unsafe archive path: " + relPosix);
            }
            return target;
        }

        public static bool IsBongoCatInstalled() => File.Exists(PluginDll) && Directory.Exists(PluginData);

        // extract the bango cat archive into the obs install root.
        public static bool InstallBongoCatPlugin(Action<string> log = null)
        {
            if (IsBongoCatInstalled())
            {
                log?.Invoke("Bongo Cat plugin already installed at " + PluginDll);
                return true;
            }
            if (!Directory.Exists(Config.PROGRAMFILES_OBS_DIR))
            {
                log?.Invoke("OBS install folder was not found: " + Config.PROGRAMFILES_OBS_DIR);
                return false;
            }
            if (!File.Exists(Config.BONGO_CAT_ZIP))
            {
                log?.Invoke(Path.GetFileName(Config.BONGO_CAT_ZIP) + " is not bundled in assets; skipping Bongo Cat");
                return false;
            }

            log?.Invoke("extracting " + Path.GetFileName(Config.BONGO_CAT_ZIP) + " -> " + Config.PROGRAMFILES_OBS_DIR);

            try
            {
                using (var zf = ZipFile.OpenRead(Config.BONGO_CAT_ZIP))
                {
                    foreach (var entry in zf.Entries)
                    {
                        string name = entry.FullName.Replace("\\", "/");
                        if (!name.StartsWith(ZipRoot)) continue;
                        string relText = name.Substring(ZipRoot.Length).Trim('/');
                        if (relText.Length == 0) continue;
                        var parts = relText.Split('/');
                        if (!IsSafeRelative(parts)) throw new InvalidOperationException("unsafe archive path: " + entry.FullName);
                        string dest = TargetPath(relText);
                        bool isDir = name.EndsWith("/") && entry.Length == 0;
                        if (isDir)
                        {
                            Directory.CreateDirectory(dest);
                            continue;
                        }
                        Directory.CreateDirectory(Path.GetDirectoryName(dest));
                        using (var src = entry.Open())
                        using (var dst = File.Open(dest, FileMode.Create, FileAccess.Write))
                        {
                            src.CopyTo(dst);
                        }
                    }
                }
            }
            catch (Exception exc)
            {
                log?.Invoke("Bongo Cat extraction failed: " + exc.Message);
                return false;
            }

            if (!File.Exists(PluginDll))
            {
                log?.Invoke("Bongo Cat extraction finished but " + PluginDll + " is missing");
                return false;
            }
            if (!Directory.Exists(PluginData))
            {
                log?.Invoke("Bongo Cat extraction finished but " + PluginData + " is missing");
                return false;
            }

            log?.Invoke("installed -> " + PluginDll);
            log?.Invoke("           + " + PluginData);
            return true;
        }
    }
}
